---
title: Snarls, pangenome deconstruction, and read mapping with vg giraffe
subtitle: GET-A-PAN
author: Xian Chang & Jean Monlong
date: 06/11/2025
output:
  revealjs::revealjs_presentation:
    theme: white
    transition: none
    highlight: pygments
    background_transition: slide
    center: false
    css: style.css
    incremental: false
    slide_level: 2
    reveal_options:
      progress: true
      slideNumber: true
      controls: true
      hash: true
      navigationMode: linear
---

# Pangenome

as a **birected** graph

![](imgs/pangenome-cartoon-nopaths.svg)

## Pangenome

as a birected graph with **haplotype paths**

![](imgs/pangenome-cartoon.svg)

# Snarls, intuitively a "bubble"

![](imgs/snarl-intro-1.svg)

# Snarls, formal definition

A snarl is a **subgraph** bounded by two **node sides** that are: 

![](imgs/snarl-intro-2.svg)

# Snarls, formal definition

A snarl is a **subgraph** bounded by two **node sides** that are: 

1. **Separable**: splitting the boundary nodes separates the snarl from the graph

![](imgs/snarl-intro-3.svg)

## Snarls, formal definition

A snarl is a **subgraph** bounded by two **node sides** that are: 

1. **Separable**: splitting the boundary nodes separates the snarl from the graph
2. **Minimal**: no nodes within the snarl are separable with either boundary node side 

![](imgs/snarl-intro-4.svg)

## Snarls, formal definition

A snarl is a **subgraph** bounded by two **node sides** that are: 

1. **Separable**: splitting the boundary nodes separates the snarl from the graph
2. **Minimal**: no nodes within the snarl are separable with either boundary node side 

![](imgs/snarl-intro-3.svg)

# Chains

A run of consecutive snarls and nodes is called a **chain**.

![](imgs/snarl-intro-chain.svg)

# Snarl decomposition

Snarls and chains can be **nested** inside of each other.

![](imgs/snarl-intro-nested.svg)

# Snarls, chains, and nodes in a tree 

The nested relationship of snarls and chains is described by the **snarl tree**.

![](imgs/snarl-intro-tree.svg)

# Netgraphs

Netgraphs are a representation of snarls with their **child chains collapsed** into a single node

![](imgs/snarl-intro-netgraph.svg)


# *Deconstruct*-ing a pangenome

Enumerate **alleles for each snarl** on **a reference path**. Before: all paths.

![](imgs/deconstruct-intro-1.svg)

## *Deconstruct*-ing a pangenome

Enumerate **alleles for each snarl** on **a reference path**. Now: only haplotypes.

![](imgs/deconstruct-intro-2.svg)

## *Deconstruct*-ing a pangenome

Enumerate **alleles for each snarl** on **a reference path** inc. nested snarls.

![](imgs/deconstruct-intro-3.svg)

# Snarl examples (`vg deconstruct`)

![](imgs/simple_nested_snarl.summary.svg)

```{.smcode}
##INFO=<ID=LV,Number=1,Type=Integer,Description="Level in the snarl tree (0=top level)">
##INFO=<ID=AT,Number=R,Type=String,Description="Allele Traversal as path in graph">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample1
ref	6	>1>4	GGCAC	CTTAG	60	.	AT=>1>2>4,>1>3>4;LV=0	GT	0|1
ref	15	>4>9	CCCAGG	CCGGTAACTACCGTCACCAGG,CCGGTACGTCA	60	.	AT=>4>8>9,>4>5>6>7>8>9,>4>5>7>9;LV=0	GT	1|2
```

# Snarl examples (`vg deconstruct`)

![](imgs/aaa.summary.svg)

```{.smcode}
##INFO=<ID=LV,Number=1,Type=Integer,Description="Level in the snarl tree (0=top level)">
##INFO=<ID=AT,Number=R,Type=String,Description="Allele Traversal as path in graph">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample1
ref	10	>1>7	AAA	AAAAAA,AAAA	60	.	AT=>1>5>6>7,>1>2>3>4>5>6>7,>1>4>5>6>7;LV=0	GT	1|2
```

## Snarl examples (`vg deconstruct`)

![](imgs/loopy_chain.summary.svg)

```{.smcode}
##INFO=<ID=LV,Number=1,Type=Integer,Description="Level in the snarl tree (0=top level)">
##INFO=<ID=AT,Number=R,Type=String,Description="Allele Traversal as path in graph">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample1	sample2
ref	11	>2>5	CTTAG	AAGTC	60	.	AT=>2>3>5,>2>4>5;LV=0	GT	0|0	1|.
```

## Snarl examples (`vg deconstruct -a`)

![](imgs/loopy_chain_flanked.summary.svg)

```{.smcode}
##INFO=<ID=LV,Number=1,Type=Integer,Description="Level in the snarl tree (0=top level)">
##INFO=<ID=PS,Number=1,Type=String,Description="ID of variant corresponding to parent snarl">
##INFO=<ID=AT,Number=R,Type=String,Description="Allele Traversal as path in graph">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	sample1	sample2
ref	13	>1>6	AGGCACCTTAGCGGTAGCTTAGCATCAG	AGGCACAAGTCCGGTAGCTTAGCATCAG,A	60	.	AT=>1>2>3>5>6,>1>2>4>5>6,>1>6;LV=0	GT	0|0	1|2
ref	19	>2>5	CTTAG	AAGTC	60	.	AT=>2>3>5,>2>4>5;LV=1;PS=>1>6	GT	0|0	1|.
```

## Snarl examples

Trick for getting this snarl decomposition to look better (currently only for the distance index):

`vg index -j [graph.dist] -w 6`

![](imgs/loopy_chain.summary.w1.svg)


# Mapping reads with `vg giraffe`


<!-- Make two columns like in https://bookdown.org/yihui/rmarkdown-cookbook/multi-column.html -->


:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 1fr; grid-column-gap: 10px;"}

::: {}
<p align="left"> Short reads</p>

![](imgs/sr-giraffe-paper.png)
:::

::: {}
<p align="left"> Long reads</p>

![](imgs/lr-giraffe-paper.png)
:::

::::

# Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

Input: Sequencing read and reference sequence

:::

::: {}


![](imgs/mapping-overview1.svg)

:::

::::

## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}


Input: Sequencing read and reference sequence

Output: Placement of the read on the reference and, usually, the edits between the sequences

:::

::: {}


![](imgs/mapping-overview7.svg)

:::

::::
## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding

:::

::: {}


![](imgs/mapping-overview2.svg)

:::

::::

## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining

:::

::: {}


![](imgs/mapping-overview3.svg)

:::

::::

## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining

:::

::: {}


![](imgs/mapping-overview4.svg)

:::

::::

## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/mapping-overview5.svg)

:::

::::


## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/mapping-overview6.svg)

:::

::::



## Read mapping algorithm overview

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/mapping-overview7.svg)

:::

::::

## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/mapping-overview8.svg)

:::

::::

## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. **Seeding** with **Minimizer Index**
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/sr-giraffe1.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. **Clustering** with **Distance Index**
1. Alignment

:::

::: {}


![](imgs/sr-giraffe2.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. **Clustering** with **Distance Index**
1. Alignment

:::

::: {}


![](imgs/sr-giraffe3.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. **Alignment**

:::

::: {}


![](imgs/sr-giraffe4.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. **Gapless extension**

:::

::: {}


![](imgs/sr-giraffe5.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. **Gapless extension** with **GBWT**

:::

::: {}


![](imgs/sr-giraffe6.svg)

:::

::::

## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. **Gapless extension** with **GBWT**

:::

::: {}


![](imgs/sr-giraffe7.svg)

:::

::::

## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. **Gapless extension** with **GBWT**

:::

::: {}


![](imgs/sr-giraffe7.1.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. Gapless extension
1. **Gapped alignment** with graph

:::

::: {}


![](imgs/sr-giraffe8.svg)

:::

::::


## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. Gapless extension
1. **Gapped alignment** with graph

:::

::: {}


![](imgs/sr-giraffe9.svg)

:::

::::



## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. Gapless extension
1. **Gapped alignment** with graph

:::

::: {}


![](imgs/sr-giraffe9.1.svg)

:::

::::

## Short read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering
1. Gapless extension
1. Gapped alignment

:::

::: {}


![](imgs/sr-giraffe10.svg)

:::

::::


# Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/lr-giraffe1.svg)

:::

::::




# Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. **Seeding** with **Minimizer Index**
1. Clustering/chaining
1. Alignment

:::

::: {}


![](imgs/lr-giraffe-seeding.svg)

:::

::::


# Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-chaining1.svg)

:::

::::


## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-linear-chaining1.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-linear-chaining2.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-linear-chaining3.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-linear-chaining4.svg)

:::

::::
## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-chaining2.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-chaining3.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Chaining**
1. Alignment


:::

::: {}

![](imgs/lr-giraffe-chaining4.svg)

:::

::::



# Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment

:::

::: {}


![](imgs/giraffe-zips1.svg)

:::

::::


## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips2.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips3.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips4.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips5.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips6.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips7.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips8.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips9.svg)

:::

::::



## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. **Zip code tree making** with **Distance Index**
1. Chaining with **Zip code trees**
1. Alignment


:::

::: {}


![](imgs/giraffe-zips10.svg)

:::

::::


## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. Zip code tree making
1. Chaining
1. **Alignment** with graph

:::

::: {}


![](imgs/lr-giraffe-chaining5.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. Zip code tree making
1. Chaining
1. **Alignment** with graph

:::

::: {}


![](imgs/lr-giraffe-alignment1.svg)

:::

::::

## Long read giraffe algorithm

:::: {style="display: grid; align-items: start; justify-items: start; grid-template-columns: 1fr 3fr; grid-column-gap: 10px;"}

::: {}

1. Seeding 
1. Zip code tree making
1. Chaining
1. Alignment

:::

::: {}


![](imgs/lr-giraffe-final-alignment.svg)

:::

::::


# Giraffe results

On the HPRC v2 graph which is x size?

![](imgs/giraffe-memory-runtime.png){.half_wide_img}

# `vg` graph formats and indexes

**Indexes** 

::: incremental

1. `.gbwt` (Graph Burrows Wheeler Transform): haplotype paths
1. `.gg` (GBWT Graph): node sequences for a GBWT
1. `.dist` (Distance Index): snarl decomposition plus minimum distances
1. `.zipcodes`: per-node distance information used by `vg giraffe`
1. `.min` (Minimizer Index): minimizers used by `vg giraffe`
1. `.gcsa` (Generalized Compressed Suffix Array): substring index used by `vg map` and `vg mpmap`

:::

**Graphs** 

::: incremental

1. `.gbz` (GBWT + GG): the graph induced by the GBWT
1. `.hg` (/`.vg`) (HashGraph): graph format optimized for speed
1. `.pg` (/`.vg`) (PackedGraph): graph format optimized for space efficiency
1. `.xg`: older graph format
1. `.vg`: protobuf-based graph format 

:::

# Conclusion

# Useful resources

- vg wiki 
  - file types: https://github.com/vgteam/vg/wiki/File-Types
  -  snarl explainer: https://github.com/vgteam/vg/wiki/Snarls-and-chains
  - deconstruct: https://github.com/vgteam/vg/wiki/VCF-export-with-vg-deconstruct

- vg manpage: https://github.com/vgteam/vg/wiki/vg-manpage

- snarls paper doi: 10.1089/cmb.2017.0251

- short read giraffe paper doi: 10.1126/science.abg8871

- long read giraffe paper doi: 10.1101/2025.09.29.678807 


