---
title: Snarls, pangenome deconstruction, and read mapping with vg giraffe
subtitle: GET-A-PAN
author: Xian Chang & Jean Monlong
date: 06/11/2025
output:
  revealjs::revealjs_presentation:
    theme: white
    transition: none
    highlight: pygments
    background_transition: slide
    center: false
    css: style.css
    incremental: false
    slide_level: 2
    reveal_options:
      progress: true
      slideNumber: true
      controls: true
      hash: true
      navigationMode: linear
---

# Pangenome graph from assemblies

Built by aligning high-quality genomes, saved as paths through the pangenome.

![](imgs/hprc-tubemap.jpeg){.half_wide_img .shadow_img}

Human Pangenome Reference Consortium (HPRC)

<div class="cite">
Liao, Asri, Ebler, et al. Nature 2023
</div>

# Pangenome

![](imgs/pangenome-cartoon-nopaths.svg)

## With haplotype paths

![](imgs/pangenome-cartoon.svg)

# Snarls, intuitively
![](imgs/snarl1.jpg)

# Snarls, formal definition

A snarl is a subgraph bounded by two node sides that are: 

(1) Separable: splitting each node into its two node sides separates a subgraph from the graph

![](imgs/snarl2.jpg)

# Snarls, formal definition

A snarl is a subgraph bounded by two node sides that are: 

(1) Separable: splitting the node into its two node sides separates a subgraph from the graph

![](imgs/snarl3.jpg)

## Snarls, formal definition

A snarl is a subgraph bounded by two node sides that are: 

(1) Separable: splitting the node into its two node sides separates a subgraph from the graph

(2) Minimal: there are no nodes within the snarl that are separable with either boundary node side 

![](imgs/snarl4.jpg)

## Snarls, formal definition

A snarl is a subgraph bounded by two node sides that are: 

(1) Separable: splitting the node into its two node sides separates a subgraph from the graph

(2) Minimal: there are no nodes within the snarl that are separable with either boundary node side 

![](imgs/snarl5.jpg)

# Chains

A run of consecutive snarls and nodes is called a chain

![](imgs/snarl6.jpg)

## Snarl decomposition

Snarls and chains can be nested inside of each other.

The nested relationship of snarls and chains is described by the snarl tree.

![](imgs/snarl7.jpg)

## Snarl decomposition

Snarls and chains can be nested inside of each other.

The nested relationship of snarls and chains is described by the snarl tree.

![](imgs/snarl8.jpg)

## Snarl decomposition

Snarls and chains can be nested inside of each other.

The nested relationship of snarls and chains is described by the snarl tree.

![](imgs/snarl9.jpg)

## Snarl decomposition

Snarls and chains can be nested inside of each other.

The nested relationship of snarls and chains is described by the snarl tree.

![](imgs/snarl10.jpg)


## Netgraphs

Netgraphs are a representation of snarls with their child chains collapsed into a single node

![](imgs/netgraph.jpg)


# Snarl examples (`vg deconstruct`)

vcf + graph + decomposition for this graph 

![](imgs/snarl2.jpg)

## Snarl examples (`vg deconstruct`)

vcf + graph + decomposition for this graph 

![](imgs/badgraph2.jpg)

## Snarl examples (`vg deconstruct`)

vcf + graph + decomposition for this graph 

![](imgs/badgraph1.png)

## Snarl examples (`vg deconstruct`)

Trick for getting this snarl decomposition to look better (currently only for the distance index):

`vg index -j [graph.dist] -w 6`


vcf + graph + decomposition for this graph 

![](imgs/badgraph1.png)


# Mapping reads with `vg giraffe`


<!-- Make two columns like in https://bookdown.org/yihui/rmarkdown-cookbook/multi-column.html -->


:::: {style="display: flex;"}

::: {}
Short reads

![](imgs/sr-giraffe-paper.png)
:::

::: {}
Long reads

![](imgs/lr-giraffe-paper.png)
:::

::::

# Long read giraffe algorithm

:::: {style="display: flex;"}

::: {}

1. Seeding with **Minimizer Index**

:::

::: {}


![](imgs/giraffe-seeding.png)

:::

::::



# Long read giraffe algorithm

:::: {style="display: flex;"}

::: {}

1. Seeding with **Minimizer Index**

2. Zip code tree making with **Distance Index**


:::

::: {}


![](imgs/giraffe-zip-tree.png)

:::

::::


## Long read giraffe algorithm

:::: {style="display: flex;"}

::: {}

1. Seeding with **Minimizer Index**

2. Zip code tree making with **Distance Index**

3. Chaining with **Zip Code Trees**

:::

::: {}


![](imgs/giraffe-chaining.png)

:::

::::

## Long read giraffe algorithm

:::: {style="display: flex;"}

::: {}

1. Seeding with **Minimizer Index**

2. Zip code tree making with **Distance Index**

3. Chaining with **Zip Code Trees**

4. Alignment with **GBWT/graph**

:::

::: {}


![](imgs/giraffe-alignment.png)

:::

::::

# Giraffe results

On the HPRC v2 graph which is x size?

![](imgs/giraffe-memory-runtime.png)

# `vg` graph formats and indexes

**Indexes** 

::: incremental

1. `.gbwt` (Graph Burrows Wheeler Transform): haplotype paths
1. `.gg` (GBWT Graph): node sequences for a GBWT
1. `.dist` (Distance Index): snarl decomposition plus minimum distances
1. `.zipcodes`: per-node distance information used by `vg giraffe`
1. `.min` (Minimizer Index): minimizers used by `vg giraffe`
1. `.gcsa` (Generalized Compressed Suffix Array): substring index used by `vg map` and `vg mpmap`

:::

**Graphs** 

::: incremental

1. `.gbz` (GBWT + GG): the graph induced by the GBWT
1. `.hg` (/`.vg`) (HashGraph): graph format optimized for speed
1. `.pg` (/`.vg`) (PackedGraph): graph format optimized for space efficiency
1. `.xg`: older graph format
1. `.vg`: protobuf-based graph format 

:::

# Conclusion

# Useful resources

- vg wiki 
  - file types: https://github.com/vgteam/vg/wiki/File-Types
  -  snarl explainer: https://github.com/vgteam/vg/wiki/Snarls-and-chains
  - deconstruct: https://github.com/vgteam/vg/wiki/VCF-export-with-vg-deconstruct

- vg manpage: https://github.com/vgteam/vg/wiki/vg-manpage

- snarls paper doi: 10.1089/cmb.2017.0251

- short read giraffe paper doi: 10.1126/science.abg8871

- long read giraffe paper doi: 10.1101/2025.09.29.678807 


